File: src/main/java/dev/choimory/member/api/common/exception/CommonExceptionHandler.java
package dev.choimory.member.api.common.exception;

import dev.choimory.member.api.common.domain.dto.CommonValidateResult;
import dev.choimory.member.api.common.domain.response.CommonResponse;
import jakarta.validation.ConstraintViolationException;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

/**
 * API 공통 예외를 HTTP 응답으로 변환하는 핸들러입니다.
 */
@Slf4j
@RestControllerAdvice("dev.choimory")
public class CommonExceptionHandler {

    /**
     * 처리되지 않은 일반 예외를 내부 서버 오류 응답으로 변환합니다.
     *
     * @param e 일반 예외
     * @return 공통 오류 응답
     */
    @ExceptionHandler(Exception.class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    public CommonResponse<String> exception(Exception e) {
        log.error("Unhandled exception", e);
        return CommonResponse.error(
                HttpStatus.INTERNAL_SERVER_ERROR,
                HttpStatus.INTERNAL_SERVER_ERROR.getReasonPhrase(),
                e.getMessage()
        );
    }

    /**
     * RuntimeException을 내부 서버 오류 응답으로 변환합니다.
     *
     * @param e 런타임 예외
     * @return 공통 오류 응답
     */
    @ExceptionHandler(RuntimeException.class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    public CommonResponse<String> runtimeException(RuntimeException e) {
        log.error("Runtime exception", e);
        return CommonResponse.error(
                HttpStatus.INTERNAL_SERVER_ERROR,
                HttpStatus.INTERNAL_SERVER_ERROR.getReasonPhrase(),
                e.getMessage()
        );
    }

    /**
     * CommonException을 지정된 HTTP 상태 응답으로 변환합니다.
     *
     * @param e 공통 예외
     * @return 공통 오류 응답 엔티티
     */
    @ExceptionHandler(CommonException.class)
    public ResponseEntity<CommonResponse<String>> commonException(CommonException e) {
        CommonResponse<String> response = CommonResponse.<String>builder()
                .code(e.getCode() != null ? e.getCode() : e.getStatus().value())
                .name(e.getCodeName() != null ? e.getCodeName() : e.getStatus().name())
                .message(e.getMsg() != null ? e.getMsg() : e.getStatus().getReasonPhrase())
                .data(null)
                .build();
        return new ResponseEntity<>(response, e.getStatus());
    }

    /**
     * Request body validation 예외를 검증 실패 응답으로 변환합니다.
     *
     * @param e validation 예외
     * @return 검증 실패 목록 응답
     */
    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public CommonResponse<List<CommonValidateResult>> methodArgumentNotValidException(MethodArgumentNotValidException e) {
        List<CommonValidateResult> results = e.getFieldErrors().stream()
                .map(fieldError -> CommonValidateResult.builder()
                        .field(fieldError.getField())
                        .rejectValue(fieldError.getRejectedValue())
                        .message(fieldError.getDefaultMessage())
                        .build())
                .toList();

        return CommonResponse.error(
                HttpStatus.BAD_REQUEST,
                HttpStatus.BAD_REQUEST.getReasonPhrase(),
                results
        );
    }

    /**
     * Constraint validation 예외를 검증 실패 응답으로 변환합니다.
     *
     * @param e constraint validation 예외
     * @return 검증 실패 목록 응답
     */
    @ExceptionHandler(ConstraintViolationException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public CommonResponse<List<CommonValidateResult>> constraintViolationException(ConstraintViolationException e) {
        List<CommonValidateResult> results = e.getConstraintViolations().stream()
                .map(violation -> CommonValidateResult.builder()
                        .field(violation.getPropertyPath().toString())
                        .rejectValue(violation.getInvalidValue())
                        .message(violation.getMessage())
                        .build())
                .toList();

        return CommonResponse.error(
                HttpStatus.BAD_REQUEST,
                HttpStatus.BAD_REQUEST.getReasonPhrase(),
                results
        );
    }

    /**
     * Spring Security 인증 예외를 인증 실패 응답으로 변환합니다.
     *
     * @param e 인증 예외
     * @return 인증 실패 응답
     */
    @ExceptionHandler(AuthenticationException.class)
    @ResponseStatus(HttpStatus.UNAUTHORIZED)
    public CommonResponse<String> authenticateException(AuthenticationException e) {
        return CommonResponse.error(
                HttpStatus.UNAUTHORIZED,
                HttpStatus.UNAUTHORIZED.getReasonPhrase(),
                e.getMessage()
        );
    }
}

File: src/main/kotlin/dev/choimory/member/api/common/exception/CommonExceptionHandler.kt
package dev.choimory.member.api.common.exception

import dev.choimory.member.api.common.domain.dto.CommonValidateResult
import dev.choimory.member.api.common.domain.response.CommonResponse
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.validation.ConstraintViolationException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.AuthenticationException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice("dev.choimory")
class CommonExceptionHandler {
    private val log = KotlinLogging.logger { }

    @ExceptionHandler(Exception::class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    fun exception(e: Exception): CommonResponse<String> {
        log.error { "${e.message}" }
        return CommonResponse(
            HttpStatus.INTERNAL_SERVER_ERROR.value(),
            HttpStatus.INTERNAL_SERVER_ERROR.reasonPhrase,
            e.message,
        )
    }

    @ExceptionHandler(RuntimeException::class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    fun runtimeException(e: RuntimeException): CommonResponse<String> {
        log.error { "${e.message}" }
        return CommonResponse(
            HttpStatus.INTERNAL_SERVER_ERROR.value(),
            HttpStatus.INTERNAL_SERVER_ERROR.reasonPhrase,
            e.message,
        )
    }

    @ExceptionHandler(CommonException::class)
    fun commonException(e: CommonException): ResponseEntity<CommonResponse<String>> {
        return ResponseEntity(
            CommonResponse(
                e.code ?: e.status.value(),
                e.msg ?: e.status.reasonPhrase,
            ),
            e.status,
        )
    }

    // Spring validation
    @ExceptionHandler(MethodArgumentNotValidException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun methodArgumentNotValidException(e: MethodArgumentNotValidException): CommonResponse<List<CommonValidateResult>> {
        return CommonResponse(
            HttpStatus.BAD_REQUEST.value(),
            HttpStatus.BAD_REQUEST.reasonPhrase,
            e.fieldErrors.stream()
                .map { f ->
                    CommonValidateResult(
                        f.field,
                        f.rejectedValue,
                        f.defaultMessage,
                    )
                }.toList(),
        )
    }

    // Spring validation
    @ExceptionHandler(ConstraintViolationException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun constraintViolationException(e: ConstraintViolationException): CommonResponse<List<CommonValidateResult>> {
        return CommonResponse(
            HttpStatus.BAD_REQUEST.value(),
            HttpStatus.BAD_REQUEST.reasonPhrase,
            e.constraintViolations.stream()
                .map { v ->
                    CommonValidateResult(
                        v.propertyPath.toString(),
                        v.invalidValue,
                        v.message,
                    )
                }.toList(),
        )
    }

    // Spring security
    @ExceptionHandler(AuthenticationException::class)
    @ResponseStatus(HttpStatus.UNAUTHORIZED)
    fun authenticateException(e: Exception): CommonResponse<String> {
        return CommonResponse(
            HttpStatus.UNAUTHORIZED.value(),
            HttpStatus.UNAUTHORIZED.reasonPhrase,
            e.message,
        )
    }
}
